macOS: support for the foreground color attribute in marked text not working

rr-codes opened the pull request:
#### 5056494ddbe484ce708028a4de28118967041414
<pre>
macOS: support for the foreground color attribute in marked text not working
<a href="https://bugs.webkit.org/show_bug.cgi?id=252792">https://bugs.webkit.org/show_bug.cgi?id=252792</a>
rdar://105812593

Reviewed by Wenson Hsieh.

* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::shouldUseHighlightsForMarkedText):
(WebKit::compositionHighlights):
(WebKit::WebViewImpl::setMarkedText):

Canonical link: <a href="https://commits.webkit.org/260773@main">https://commits.webkit.org/260773@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/878c2ddc70ea913e09d41f8e4af4cf6da9be644d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109174 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18253 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41986 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/700 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118410 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19712 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9555 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101414 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114931 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14780 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98010 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42950 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96752 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29664 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84683 "Found 1 new API test failure: /WebKitGTK/TestDownloads:/webkit/Downloads/local-file-error (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11030 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31007 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11758 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7940 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17127 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50610 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7453 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13378 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4935,6 +4935,71 @@ static bool eventKeyCodeIsZeroOrNumLockOrFn(NSEvent *event)
+    mergedHighlights.reserveInitialCapacity(highlights.size());
+    for (auto& highlight : highlights) {
+        if (mergedHighlights.isEmpty() || mergedHighlights.last().backgroundColor != highlight.backgroundColor || mergedHighlights.last().foregroundColor != highlight.foregroundColor)
+            mergedHighlights.append(highlight);
```
Can uncheckedAppend() here.

cdumez commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4935,6 +4935,71 @@ static bool eventKeyCodeIsZeroOrNumLockOrFn(NSEvent *event)
+            mergedHighlights.last().endOffset = highlight.endOffset;
+    }
+
+    return mergedHighlights;
```
Is this vector stored somewhere for an extended period of time. It is unclear from this patch alone. If it is the case, then we should call shrinkToFit() before returning it since you may have reserved too much capacity.

cdumez commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4968,7 +5038,7 @@ static bool eventKeyCodeIsZeroOrNumLockOrFn(NSEvent *event)
     }
 
-    m_page->setCompositionAsync(text, underlines, { }, selectedRange, replacementRange);
+    m_page->setCompositionAsync(text, underlines, highlights, selectedRange, replacementRange);
```
Can WTFMove() the 2 vectors.

whsieh reviewed (approved):
Seems okay to me, with Chris's earlier review.

Since underlines and highlights are mutually exclusive, you could also consider using a single `std::variant` to formalize that notion.

rr-codes committed:
macOS: support for the foreground color attribute in marked text not working
https://bugs.webkit.org/show_bug.cgi?id=252792
rdar://105812593

Reviewed by Wenson Hsieh.

* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::shouldUseHighlightsForMarkedText):
(WebKit::compositionHighlights):
(WebKit::WebViewImpl::setMarkedText):

Canonical link: https://commits.webkit.org/260773@main
